pkg/yqlib/decoder_goccy_yaml.go

```diff
@@ -1,49 +1,238 @@
 //go:build !yq_noyaml
 
-//
-// NOTE this is still a WIP - not yet ready.
-//
-
+// Package yqlib provides YAML processing functionality using the goccy/go-yaml parser.
+// This decoder implementation provides compatibility with the legacy yaml.v3 parser
+// while leveraging the actively maintained goccy/go-yaml library.
 package yqlib
 
 import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
 	"io"
+	"regexp"
+	"strings"
 
 	yaml "github.com/goccy/go-yaml"
 	"github.com/goccy/go-yaml/ast"
 )
 
+// Constants for document processing
+const (
+	// docSeparatorPlaceholder is used internally to mark document separators during preprocessing.
+	docSeparatorPlaceholder = "$yqDocSeparator$"
+)
+
 type goccyYamlDecoder struct {
 	decoder yaml.Decoder
 	cm      yaml.CommentMap
-	// anchor map persists over multiple documents for convenience.
+
+	prefs YamlPreferences
+
+	// leadingContent stores comments and directives found before the actual YAML content.
+	leadingContent string
+	// bufferRead is used with a TeeReader when LeadingContentPreProcessing is off,
+	// to capture initial bytes that might only contain comments.
+	bufferRead bytes.Buffer
+
+	// anchorMap stores anchors encountered within the current YAML stream.
+	// It is reset by Init() for each new stream.
 	anchorMap map[string]*CandidateNode
+
+	// dateTimePreprocessor handles automatic timestamp tagging for datetime arithmetic compatibility
+	dateTimePreprocessor *DateTimePreprocessor
+
+	readAnything  bool // Flag to track if any actual YAML node (or synthesized comment node) has been decoded.
+	firstFile     bool // Flag for 'evaluateTogether' mode to handle leading content only once.
+	documentIndex uint // Index of the current document within the stream.
+}
+
+// NewGoccyYAMLDecoder creates a new YAML decoder using the goccy/go-yaml library,
+// configured with the given preferences.
+func NewGoccyYAMLDecoder(prefs YamlPreferences) Decoder {
+	return &goccyYamlDecoder{
+		prefs:                prefs,
+		firstFile:            true,
+		dateTimePreprocessor: NewDateTimePreprocessor(true), // Enable datetime preprocessing for Goccy
+	}
 }
 
-func NewGoccyYAMLDecoder() Decoder {
-	return &goccyYamlDecoder{}
+// processReadStream pre-processes the input stream to extract leading comments,
+// directives, and document separators before the main YAML content is parsed.
+// It returns a reader for the remaining stream and the extracted leading content.
+func (dec *goccyYamlDecoder) processReadStream(reader *bufio.Reader) (io.Reader, string, error) {
+	var commentLineRegEx = regexp.MustCompile(`^\s*#`)
+	var yamlDirectiveLineRegEx = regexp.MustCompile(`^\s*%YA`) // e.g., %YAML
```

Comment:
Move this out, otherwise the regexp is computed each time the function is called